### PR TITLE
Add Cilium regression tests

### LINKED ISSUE #832: Add regression test for ebpf-for-windows-demo
### Describe the feature you'd like supported

Verify we don't have a breaking change in the verifier (at least) that breaks the demo repository by running some test on each PR.

### Proposed solution

There are three potential directions:
a) Put the demo code into the main ebpf-for-windows repo.  I'd be reluctant to do this as it would increase build times and disk space for everyone who tries to clone and build the repo.
b) Download the relevant artifacts from the demo repo's build artifacts and use them in a test in the main ebpf-for-windows repo, such as a "show verification" test.   This may be the best option.
c) Add dependabot to the demo repo, and make sure the CI/CD there runs tests on the demo repo.   This would cause any failure to show up in the demo repo rather than being a check in the main repo.

My current inclination is to use approach B for verifier tests, and approach C for any more complex tests.

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by saxena-anurag**
## Description

This PR does the following:
1. Fix a bug in `bpf_object__open_file()`
2. Download artifacts from `ebpf-for-windows-demo` repo.
3. Run verification tests on Cilium .o files in CICD. Verification happens only for Release build (as Debug build takes a lot of time)

## Testing

Added new tests.

## Documentation

NA

Fixes #832

### CODE REVIEW

File: .github/workflows/reusable-build.yml [hunk 1]
```diff
@@ -98,6 +98,25 @@ jobs:
       working-directory: ./${{env.BUILD_PLATFORM}}/${{env.BUILD_CONFIGURATION}}
       run: copy "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\clang*" .
 
+    - name: Download demo repository artifacts
+      uses: dawidd6/action-download-artifact@v2.19.0
+      with:
```
Comment (Alan-Jowett) [inline .github/workflows/reusable-build.yml:103]: Why is this needed during the build and not during the test phase?

File: .github/workflows/reusable-build.yml [hunk 2]
```diff
@@ -98,6 +98,25 @@ jobs:
       working-directory: ./${{env.BUILD_PLATFORM}}/${{env.BUILD_CONFIGURATION}}
       run: copy "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\clang*" .
 
+    - name: Download demo repository artifacts
+      uses: dawidd6/action-download-artifact@v2.19.0
```
Comment (Alan-Jowett) [inline .github/workflows/reusable-build.yml:102]: Can you pin this to a specific commit id?

File: tests/cilium/cilium_tests.cpp [hunk 3]
```diff
@@ -0,0 +1,59 @@
+// Copyright (c) Microsoft Corporation
+// SPDX-License-Identifier: MIT
+
+#define CATCH_CONFIG_MAIN
+
+#include <iostream>
+
+#include "ebpf_api.h"
+#include "bpf/libbpf.h"
```
Comment (dthaler) [inline tests/cilium/cilium_tests.cpp:9]: Sort in alphabetical order?

File: tests/cilium/cilium_tests.cpp [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+#include "catch_wrapper.hpp"
+
+#define SAMPLE_PATH ""
+#define CILIUM_XDP_SECTIONS_SNAT 10
+#define CILIUM_XDP_SECTIONS_DSR 12
+
+void
+verify_program(_In_ const char* file, uint32_t expected_section_count)
```
Comment (dthaler) [inline tests/cilium/cilium_tests.cpp:17]:
```suggestion
verify_program(_In_z_ const char* file, uint32_t expected_section_count)
```

File: .github/workflows/reusable-build.yml (hunk 1 above)
Comment (saxena-anurag) [reply .github/workflows/reusable-build.yml:103]: I added it during the build phase so that the downloaded artifacts are then uploaded along with the build artifacts. Then:
1. The tests will directly get these when they download the build artifacts
2. If a test fails, the .o files used for the test will be available in the uploaded build artifacts.

Comment (dthaler) [reply .github/workflows/reusable-build.yml:103]: Suggest adding # comments in the file to explain.

Comment (saxena-anurag) [reply .github/workflows/reusable-build.yml:106]: Added.

File: .github/workflows/reusable-build.yml [hunk 5]
```diff
@@ -101,6 +101,28 @@ jobs:
       working-directory: ./${{env.BUILD_PLATFORM}}/${{env.BUILD_CONFIGURATION}}
       run: copy "C:\Program Files (x86)\Microsoft Visual Studio\2019\Enterprise\VC\Tools\MSVC\14.29.30133\bin\Hostx64\x64\clang*" .
 
+    # Download and extract demo repo artifacts. These then get uploaded along with the build artifacts.
+    # The subequent tests will directly get these when they download the build artifacts. In case
```
Comment (dthaler) [inline .github/workflows/reusable-build.yml:105]:
```suggestion
    # The subsequent tests will directly get these when they download the build artifacts. In case
```

**Review by dthaler — APPROVED** [review]

**Review by Alan-Jowett — APPROVED** [review]
